### PR TITLE
feat: support hashbang interpreter comments

### LINKED ISSUE #429: feat: interpreterDirective (a.k.a `#!shebang`)
Oxc currently parses shebang as a comment, but we should change it to a dedicated ast node.

This will make the formatter / printer a little easier.

Steps:
* inside the lexer, modify 
https://github.com/Boshen/oxc/blob/9df154560298bb754efcd33aded89618e1490794/crates/oxc_parser/src/lexer/mod.rs#L423-L427
and let it return a dedicated token
* inside the ast, add a dedicated ast node to `Program`, see https://github.com/babel/babel/blob/d062d47d57a679953144ccf1e60b3efc297e5b43/packages/babel-parser/src/types.d.ts#L182
* inside the parser, consume the token and add the node to `Program`
* inside hir and ast_lower, copy this ast node to HIR
* inside the minifier, print the hashbang
* this test should pass in the end

https://github.com/Boshen/oxc/blob/9df154560298bb754efcd33aded89618e1490794/crates/oxc_minifier/tests/tdewolff/mod.rs#L7

### PR DESCRIPTION
**Pull request by sno2**
Closes #429

### CODE REVIEW

File: crates/oxc_ast/src/ast/js.rs [hunk 1]
```diff
@@ -19,6 +19,7 @@ pub struct Program<'a> {
     pub span: Span,
     pub source_type: SourceType,
     pub directives: Vec<'a, Directive<'a>>,
+    pub interpreter: Option<Hashbang<'a>>,
```
Comment (Boshen) [inline crates/oxc_ast/src/ast/js.rs:22]: Shall we change the key to `shashbang` as well? I've got no idea what a `interpreter` (from babel). The spec just calls it a hashbang.

File: crates/oxc_minifier/src/printer/gen.rs [hunk 2]
```diff
@@ -50,6 +50,14 @@ impl<'a> Gen for Program<'a> {
     }
 }
 
+impl <'a> Gen for Hashbang<'a> {
+	fn gen(&self, p: &mut Printer, ctx: Context) {
+		p.print(b'#');
+		p.print(b'!');
```
Comment (Boshen) [inline crates/oxc_minifier/src/printer/gen.rs]: You can print them together with `print_str`

File: crates/oxc_parser/src/lexer/mod.rs [hunk 3]
```diff
@@ -420,11 +420,10 @@ impl<'a> Lexer<'a> {
             }
             '^' => self.read_caret(),
             '#' => {
-                // https://tc39.es/proposal-hashbang/out.html
```
Comment (sno2) [inline crates/oxc_parser/src/lexer/mod.rs:423]: it's part of the spec now https://tc39.es/ecma262/multipage/ecmascript-language-lexical-grammar.html#sec-hashbang

File: crates/oxc_minifier/tests/tdewolff/mod.rs [hunk 4]
```diff
@@ -4,7 +4,7 @@ use crate::test;
 
 #[test]
 fn tdewolff() {
-    // expect("#!shebang", "#!shebang");
+    test("#!shebang", "#!shebang");
```
Comment (Boshen) [inline crates/oxc_minifier/tests/tdewolff/mod.rs:7]: Nice!

File: crates/oxc_parser/src/lexer/mod.rs [hunk 5]
```diff
@@ -514,6 +513,17 @@ impl<'a> Lexer<'a> {
         Kind::MultiLineComment
     }
 
+    /// Section 12.5 Hashbang Comments
+    fn skip_shebang_comment(&mut self) -> Kind {
```
Comment (Boshen) [inline crates/oxc_parser/src/lexer/mod.rs]: nit: `read_shebang_comment`, we are not skipping it.

File: crates/oxc_ast/src/ast/js.rs (hunk 1 above)
Comment (sno2) [reply crates/oxc_ast/src/ast/js.rs]: yeah, I thought of that at the beginning but wasn't sure

**Comment by Boshen** [comment]
That was fast! Thank you so much.

**Pull request merged**